src/modules/kubernetes.rs

```diff
@@ -80,30 +86,44 @@ fn get_kube_ctx_component(filename: path::PathBuf, current_ctx: &str) -> Option<
     Some(ctx_components)
 }
 
-fn get_kube_user<'a>(config: &'a KubernetesConfig, kube_user: &'a str) -> Cow<'a, str> {
-    return get_alias(&config.user_aliases, kube_user).unwrap_or(Cow::Borrowed(kube_user));
-}
-
-fn get_kube_context_name<'a>(config: &'a KubernetesConfig, kube_ctx: &'a str) -> Cow<'a, str> {
-    return get_alias(&config.context_aliases, kube_ctx).unwrap_or(Cow::Borrowed(kube_ctx));
-}
-
-fn get_alias<'a>(
-    aliases: &'a HashMap<String, &'a str>,
-    alias_candidate: &'a str,
-) -> Option<Cow<'a, str>> {
-    if let Some(val) = aliases.get(alias_candidate) {
-        return Some(Cow::Borrowed(val));
+fn get_aliased_name<'a>(
```

Comment:
I suspect this is partially because I'm tired, but I've had a hell of a time trying to parse how this function is used. I've been staring at this and its call sites for almost 20 minutes now and there are still things I don't understand:

- Why do we take two `Option<&str>` and then an `Option<&String>`?
- What do these strings represent? (Every time I think I have it, one of them falls out of my head and I have to start thinking about it again).
- Why does the one of the call sites have a `Some(Some(something, _, _))` followed immediately by an `unwrap_or_else`?

I'm afraid I don't have a concrete suggestion right now (maybe will have one this weekend) but my gut is that this is too complex to merge as it is. Perhaps there's no better way to write it, in which case I'll grumble a little and approve, but I suspect there's a way to make this simpler than it currently is.

Reply:
I think it's more likely due to my inexperience with rust :-/

> Why do we take two Option<&str> and then an Option<&String>?

Mostly because the inputs were this and this version compiled :-/ I still haven't completely wrapped my head around when to use &str vs String vs &String...

> What do these strings represent? (Every time I think I have it, one of them falls out of my head and I have to start thinking about it again).

I think of this function as a wrapper around a regex search and replace: have a pattern, have a current value, have a replacement. The alias is the replacement. Only difference: if the pattern doesn't match, return None instead of current value.

> Why does the one of the call sites have a Some(Some(something, _, _)) followed immediately by an unwrap_or_else?

Again, probably because I got it compiled like that when using the filter_map (if you mean line 201) :-(

Just to clarify, the part around line 201:

```
   let (matched_context_config, display_context, display_user) = config.contexts.iter().find_map(|context_config| {
        let Some(context_alias) = get_aliased_name(Some(context_config.context_pattern), Some(&current_kube_ctx_name), context_config.context_alias) else {
            return None;
        };

        let user_alias = ... 
...
        Some((Some(context_config), context_alias, user_alias))
    }).unwrap_or_else(|| (None, current_kube_ctx_name.clone(), ctx_components.user));

```

So `Some(Some(something, _, _))` is actually `Some(Some(something), _, _)`,, so `Option<Tuple<Option<..>,..>> which hopefully is a bit more sane than an `Option<Option<Tuple>>`

>  but I suspect there's a way to make this simpler than it currently is.

I sure hope, but I'm currently out of my wits/experience for anything to make this easier. :-(

Reply:
Ok, came across this: https://rust-unofficial.github.io/patterns/idioms/coercion-arguments.html which argues that you should always go with `&str` instead of `&String`. Will make the adjustments after I understood that chapter...

Reply:
Did some refinement on the function... Hope it's clearer now. Also got rid of the `&String`